botocore/endpoint.py

```diff
@@ -199,24 +199,24 @@ def _get_response(self, request, operation_model, attempts):
                                                  operation_model)
         parser = self._response_parser_factory.create_parser(
             operation_model.metadata['protocol'])
-        return ((http_response,
-                 parser.parse(response_dict, operation_model.output_shape)),
-                None)
+        parsed_response = parser.parse(
+            response_dict, operation_model.output_shape)
+        return (http_response, parsed_response), None
 
     def _looks_like_dns_error(self, e):
         return 'gaierror' in str(e) and e.request is not None
 
     def _looks_like_bad_status_line(self, e):
         return 'BadStatusLine' in str(e) and e.request is not None
 
-    def _needs_retry(self, attempts, operation_model, response=None,
-                     caught_exception=None):
+    def _needs_retry(self, attempts, operation_model, request_dict,
+                     response=None, caught_exception=None):
         event_name = 'needs-retry.%s.%s' % (self._endpoint_prefix,
                                             operation_model.name)
         responses = self._event_emitter.emit(
             event_name, response=response, endpoint=self,
             operation=operation_model, attempts=attempts,
-            caught_exception=caught_exception)
+            caught_exception=caught_exception, request_dict=request_dict)
```

Comment:
Initially I was inclined to say that let's avoid adding a kwarg if possible.  However, I'm not opposed to adding new kwargs if it exposes new general purpose functionality.

The way I see it, we need the general ability to be able to not only indicate that a request should be retried, but allow handlers to alter the parameters of subsequent requests.

I think Jordon's approach here does that.  It opens up the possibility for handlers to change how requests are retried by modifying the `request_dict`.  I think that's pretty reasonable case for adding a new kwarg.